consensus/reactor.go

```diff
@@ -494,16 +494,16 @@ OUTER_LOOP:
 			}
 		}
 
-		// If the peer is on a previous height, help catch up.
-		if (0 < prs.Height) && (prs.Height < rs.Height) {
+		// If the peer is on a previous height that we have, help catch up.
+		if (0 < prs.Height) && (prs.Height < rs.Height) && (prs.Height >= conR.conS.blockStore.Base()) {
```

Comment:
Is it possible for pruning to race this? So this condition might be valid but by the time we try to execute LoadBlockMeta below we've pruned it? I guess that's what all the nil checks help catch but maybe we want an atomic Base+Load

Reply:
Funny timing: https://github.com/tendermint/tendermint/issues/5377

Reply:
Yeah, the `nil` check will handle that as well as other failure scenarios. An atomic base+load won't help here, since the peer may be requesting the block at height base+20 concurrently with base to base+100 being pruned - we'll have to check for `nil`. The base check is just to avoid having to query the database at all if we know it can't be satisfied.